### PR TITLE
Use FloorDiv and Mod instead of // and % on sympy exprs

### PR DESCRIPTION
**Pull request by shino16**
Fixes #162868. Continuation of #175755.

Inductor call sites use Python `%` and `//` on raw sympy symbols, which create `sympy.Mod` and `sympy.floor` -- different types from the `torch.utils._sympy.functions.Mod`/`FloorDiv` that `torch._check` installs as axioms. This causes `statically_known_true` to miss divisibility constraints, leading to unnecessary masking and missed optimizations in generated Triton kernels.

16 call sites fixed across 8 files. ~900 other occurrences of `//`/`%` were triaged and confirmed safe (SymInt dispatch, concrete values, codegen output, or intentional sympy.Mod usage in the congruence solver).

## Benchmark: flex_attention 2.5x regression with dynamic shapes

The most impactful fix is in flex_attention, where `seq_len_q % 128 != 0` always evaluated to `True` for symbolic shapes (bypassing axioms entirely), forcing `IS_DIVISIBLE=False` unconditionally. Benchmark on GB200, `flex_attention` with S=2048 and `torch._check(seq_len % 128 == 0)`:

| | Before | After |
|---|---|---|
| `dynamic=False` | 0.43 ms | 0.43 ms |
| `dynamic=True` | 1.07 ms (2.5x regression) | 0.43 ms |

```python
import torch
import torch.utils.benchmark as benchmark
from torch.nn.attention.flex_attention import flex_attention

B, H, S, D = 4, 32, 2048, 64

for dynamic in [False, True]:
    torch._dynamo.reset()
    def fn(q, k, v):
        torch._check(q.shape[2] % 128 == 0)
        torch._check(k.shape[2] % 128 == 0)
        return flex_attention(q, k, v)
    compiled = torch.compile(fn, dynamic=dynamic)
    q = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)
    k = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)
    v = torch.randn(B, H, S, D, device="cuda", dtype=torch.float16)
    for _ in range(3): compiled(q, k, v)
    torch.cuda.synchronize()
    t = benchmark.Timer(
        stmt="fn(q, k, v)", globals={"fn": compiled, "q": q, "k": k, "v": v},
    )
    print(f"dynamic={str(dynamic):5s}: {t.blocked_autorange(min_run_time=0.5).median*1e3:.4f} ms")
```

See also [#175752](https://github.com/pytorch/pytorch/issues/175752) for the RMSNorm vectorization regression (27% on H100) fixed by [#175755](https://github.com/pytorch/pytorch/pull/175755).

### Scope

Searched **444 Python files** across the compiler codebase:

- `torch/_inductor/` (the main inductor compiler)
- `torch/_dynamo/` (the graph capture frontend)
- `torch/fx/experimental/symbolic_shapes.py` (core shape inference)
- `torch/utils/_sympy/` (sympy utilities: functions, reference, value_ranges, printers)

It was done against PyTorch commit e47de660122.

### Method

Performed two parallel exhaustive searches:

1. **All `//` uses** on potentially-sympy operands across all 444 files
2. **All `%` uses** on potentially-sympy operands across all 444 files, excluding string formatting, comments, and codegen output

Out of **917 raw occurrences** of `//` or `%` across 97 files, each was triaged by examining the operand types in context:

- Is the operand a `sympy.Expr` (bare sympy expression)? → Needs fix
- Is the operand a `SymInt`? → Safe (`SymInt.__mod__`/`__floordiv__` dispatch through symnode which produces `Mod`/`FloorDiv` correctly)
- Is the operand a Python `int` / `sympy.Integer` (concrete)? → Safe (concrete arithmetic gives correct results)
- Is it in a string, comment, codegen template, or Jinja output? → Not applicable

### Results

| Category | Count | Action |
|---|---|---|
| **Fixed (sympy.Expr operands)** | 16 sites across 8 files | Changed to `Mod`/`FloorDiv` |
| **Excluded — SymInt dispatch** | ~10 sites | Safe: `SymInt` operators produce correct types |
| **Excluded — concrete values** | ~15 sites | Safe: Python int or `sympy.Integer` arithmetic |
| **Excluded — codegen/templates** | ~50+ sites | Not Python execution, just output strings |
| **Excluded — string formatting** | ~100+ sites | `%s`, `% (`, etc. |
| **Excluded — comments/URLs** | ~700+ sites | Not code |
| **Excluded — internal sympy solver** | 4 sites in `symbolic_shapes.py:3080-3236` | Congruence system feeds into sympy's equation solver which expects `sympy.Mod` |
| **Excluded — value range analysis** | 2 sites in `value_ranges.py` | Operates on concrete bounds |

## Changes made

Three distinct patterns:

### Pattern A: Axiom mismatch via `xreplace`

`statically_known_true(Eq(s % N, 0))` creates `sympy.Mod(s, N)`. The axiom uses `torch.utils._sympy.functions.Mod(s, N)`. They print identically but `xreplace` requires exact type match.

### Pattern B: Direct comparison bypasses axiom system

`s % 128 != 0` on a symbolic `sympy.Expr` resolves to concrete `True` (sympy's `__ne__`), unconditionally taking the "not divisible" branch. The fix replaces this with `statically_known_true(Eq(Mod(s, 128), 0))` which consults axioms.

### Pattern C: `//` creates wrong expression tree

`s // t` creates `floor(s/t)`, not `FloorDiv(s, t)`. Axioms and pattern matching on `FloorDiv` don't find the `floor(...)` form.

| File | Lines | Change | Bug pattern |
|---|---|---|---|
| `sizevars.py` | 941 | `mod2 % mod` → `Mod(mod2, mod)` | A |
| `ir.py` | 1858, 2329 | `reduction_numel % split` → `Mod(...)` in `sympy.Eq` | A |
| `lowering.py` | 3045, 3050 | `stride % ALIGNMENT` → `Mod(...)` | A |
| `loop_body.py` | 272 | `iter_idx[dim] % original_range` → `Mod(...)` | A, C |
| `flex_attention.py` | 321-328 | Rewrote direct `% 128 != 0` to `statically_known_true(Eq(Mod(...), 0))` | A, B |
| `flex_attention.py` | 356, 868 | `Hq // Hkv` → `FloorDiv(Hq, Hkv)` | C |
| `flex_attention.py` | 710-713 | `seq_len % 128 == 0` → `Eq(Mod(...), 0)` | A |
| `flex_decoding.py` | 72, 195 | `Hq // Hkv` → `FloorDiv(Hq, Hkv)` | C |
| `flex_decoding.py` | 184, 187 | `seq_len % 128` → `Mod(...)` in `sympy.Eq` | A |
| `flex_decoding.py` | 305 | `(expr) % BLOCK_M` → `Mod(...)` | A, B |
| `triton.py` | 2043 | `k % BLOCK_K` → `Mod(...)` in `sympy.Eq` | A |
| `symbolic_shapes.py` | 6874 | `base % divisor` → `Mod(base, divisor)` | B |

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by laithsakka — APPROVED** [review]
Thank you !!!
